django/contrib/auth/views.py

```diff
@@ -438,20 +442,35 @@ class PasswordResetConfirmView(PasswordContextMixin, FormView):
     title = _('Enter new password')
     token_generator = default_token_generator
 
+    def _build_redirect_url(self):
+        current_token = self.request.resolver_match.kwargs['token']
+        return self.request.path.replace(current_token, INTERNAL_RESET_URL_TOKEN)
+
     @method_decorator(sensitive_post_parameters())
     @method_decorator(never_cache)
     def dispatch(self, *args, **kwargs):
         assert 'uidb64' in kwargs and 'token' in kwargs
 
         self.validlink = False
-        self.user = self.get_user(kwargs['uidb64'])
 
-        if self.user is not None and self.token_generator.check_token(self.user, kwargs['token']):
+        # Store the token in the session, to not leak it in a referer
+        # and redirect to a token-free URL.
+        # See https://robots.thoughtbot.com/is-your-site-leaking-password-reset-links for details.
+        token = kwargs['token']
+        if token != INTERNAL_RESET_URL_TOKEN:
```

Comment:
@collinanderson That is a good point, at first I wanted to be safe against leaking anything (just to be on the safe side), but the DOS argument is more important.